### PR TITLE
Add KNNImputer

### PR DESCRIPTION
**Pull request by srzeszut**
I have added the KNNImputer and I am currently implementing tests to ensure that it behaves as expected across various scenarios, including edge cases.

### CODE REVIEW

File: lib/scholar/impute/knn_imputer.ex [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+    The function returns a struct with the following parameters:
+
+    * `:missing_values` - the same value as in `:missing_values`
+
+    * `:statistics` - The imputation fill value for each feature. Computing statistics can result in
+    [`Nx.Constant.nan/0`](https://hexdocs.pm/nx/Nx.Constants.html#nan/0) values.
+
+  ## Examples
+  iex> x = Nx.tensor([[40.0, 2.0],[4.0, 5.0],[7.0, :nan],[:nan, 8.0],[11.0, 11.0]])
```
Comment (josevalim) [inline lib/scholar/impute/knn_imputer.ex:49]: Remember to indent docs example with 4 spaces from the heredoc. So you must add four spaces here :)

File: lib/scholar/impute/knn_imputer.ex [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    if input_rank != 2 do
+      raise ArgumentError, "Wrong input rank. Expected: 2, got: #{inspect(input_rank)}"
+    end
+
+    if opts[:missing_values] != :nan and
+         Nx.any(Nx.is_nan(x)) == Nx.tensor(1, type: :u8) do
+      raise ArgumentError,
+            ":missing_values other than :nan possible only if there is no Nx.Constant.nan() in the array"
+    end
+
```
Comment (josevalim) [inline lib/scholar/impute/knn_imputer.ex:80]: This check does not really work in Nx. If you call `fit` inside `Nx.Defn.jit`, then `x` is an expression, and we can't read its values to find out if there is a nan or not. The best we can do is to remove this check and document it.

File: lib/scholar/impute/knn_imputer.ex [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    {rows, cols} = Nx.shape(x)
+
+    row_nan_count = Nx.sum(Nx.is_nan(x), axes: [1])
+    # row with only 1 non nan value is also considered as all nan row
+    all_nan_rows =
+      Nx.select(Nx.greater_equal(row_nan_count, cols - 1), Nx.tensor(1), Nx.tensor(0))
+
+    all_nan_rows_count = Nx.sum(all_nan_rows)
+
+    if num_neighbors > rows - 1 - Nx.to_number(all_nan_rows_count) do
```
Comment (josevalim) [inline lib/scholar/impute/knn_imputer.ex:101]: Same here, this code won't work because, when you have an expression, you can't get a number from it. Can we remove this check? What happens if we don't check for this condition?

Comment (josevalim) [reply lib/scholar/impute/knn_imputer.ex:101]: You can test this by calling `fit` after jitting it with `Nx.Defn.fit`.

File: lib/scholar/impute/knn_imputer.ex [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+      raise ArgumentError,
+            "Number of neighbors rows must be less than number valid of rows - 1 (valid row is row with more than 1 non nan value)"
+    end
+
+    placeholder_value = Nx.Constants.nan() |> Nx.tensor()
+
+    statistics = knn_impute(x, placeholder_value, num_neighbors: num_neighbors)
```
Comment (josevalim) [inline lib/scholar/impute/knn_imputer.ex:108]: You should be fine to pass it as a regular argument, since it is always a number?

```suggestion
    statistics = knn_impute(x, placeholder_value, num_neighbors)
```

File: lib/scholar/impute/knn_imputer.ex [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+      while {{row = 0, mask, num_neighbors, num_rows, x}, values_to_impute},
+            Nx.less(row, num_rows) do
+        {_, values_to_impute} =
+          while {{col = 0, mask, num_neighbors, num_cols, row, x}, values_to_impute},
+                Nx.less(col, num_cols) do
+            if mask[row][col] > 0 do
+              {rows, cols} = Nx.shape(x)
+
+              neighbor_avg =
+                calculate_knn(x, row, col, rows: rows, num_neighbors: opts[:num_neighbors])
```
Comment (josevalim) [inline lib/scholar/impute/knn_imputer.ex:158]:
```suggestion
                calculate_knn(x, row, col, rows, num_neighbors)
```

File: lib/scholar/impute/knn_imputer.ex [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+    # minus nan column
+    coordinates = coordinates - 1
+
+    # inputes zeros in nan_col to calculate distance with squared_euclidean
+    new_row = Nx.indexed_put(row, Nx.new_axis(nan_col, 0), Nx.tensor(0))
+
+    # if potential neighbor has nan in nan_col, we don't want to calculate distance and the case if potential_neighbour is the row to impute
+    {potential_neighbor} =
+      if potential_neighbor[nan_col] == Nx.Constants.nan() do
```
Comment (josevalim) [inline lib/scholar/impute/knn_imputer.ex:224]: I am not sure if this check is guaranteed to work, given two NaNs are not guaranteed to be equal. Using `Nx.is_nan` would be more appropriate.

File: lib/scholar/impute/knn_imputer.ex [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+  """
+  import Nx.Defn
+  import Scholar.Metrics.Distance
+
+  @derive {Nx.Container, keep: [:missing_values], containers: [:statistics]}
+  defstruct [:statistics, :missing_values]
+
+  opts_schema = [
+    missing_values: [
+      type: {:or, [:float, :integer, {:in, [:nan]}]},
```
Comment (msluszniak) [inline lib/scholar/impute/knn_imputer.ex:17]: Why `:nan` here is wrapped inside `:in`?

File: lib/scholar/impute/knn_imputer.ex [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+    if opts[:missing_values] != :nan and
+         Nx.any(Nx.is_nan(x)) == Nx.tensor(1, type: :u8) do
+      raise ArgumentError,
+            ":missing_values other than :nan possible only if there is no Nx.Constant.nan() in the array"
+    end
+
+    x =
+      if opts[:missing_values] != :nan,
+        do: Nx.select(Nx.equal(x, opts[:missing_values]), Nx.Constants.nan(), x),
```
Comment (msluszniak) [inline lib/scholar/impute/knn_imputer.ex:83]: Use `Nx.is_nan` here NaN is not equal to itself

File: lib/scholar/impute/knn_imputer.ex [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+    if num_neighbors < 1 do
+      raise ArgumentError, "Number of neighbors must be greater than 0"
+    end
+
+    {rows, cols} = Nx.shape(x)
+
+    row_nan_count = Nx.sum(Nx.is_nan(x), axes: [1])
+    # row with only 1 non nan value is also considered as all nan row
+    all_nan_rows =
+      Nx.select(Nx.greater_equal(row_nan_count, cols - 1), Nx.tensor(1), Nx.tensor(0))
```
Comment (msluszniak) [inline lib/scholar/impute/knn_imputer.ex:97]: From the check in 88th line, everything might be moved to defn as you cannot effectively execute the check below. Then, this can be simplify to `all_nan_rows = row_nan_count <= cols - 1)`

File: lib/scholar/impute/knn_imputer.ex [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+
+        if i == nan_row do
+          distance = Nx.Constants.infinity({:f, 32})
+          row_distances = Nx.indexed_put(row_distances, Nx.new_axis(i, 0), distance)
+          {{i + 1, x, row_with_value_to_fill, rows, nan_row, nan_col}, row_distances}
+        else
+          distance = nan_euclidian(row_with_value_to_fill, nan_col, potential_donor)
+          row_distances = Nx.indexed_put(row_distances, Nx.new_axis(i, 0), distance)
+          {{i + 1, x, row_with_value_to_fill, rows, nan_row, nan_col}, row_distances}
+        end
```
Comment (msluszniak) [inline lib/scholar/impute/knn_imputer.ex:202]:
```suggestion
        distance = if i == nan_row do
          Nx.Constants.infinity({:f, 32})
        else
          nan_euclidian(row_with_value_to_fill, nan_col, potential_donor)
        end
        row_distances = Nx.indexed_put(row_distances, Nx.new_axis(i, 0), distance)
        {{i + 1, x, row_with_value_to_fill, rows, nan_row, nan_col}, row_distances}
```

File: lib/scholar/impute/knn_imputer.ex [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+
+        {potential_neighbor}
+      end
+
+    # calculates how many values are present in the row without nan_col to calculate weight for the distance
+    present_coordinates = Nx.sum(Nx.logical_not(Nx.is_nan(potential_neighbor))) - 1
+
+    # if row has all nans we skip it
+    {weight, potential_neighbor} =
+      if present_coordinates == 0 do
```
Comment (msluszniak) [inline lib/scholar/impute/knn_imputer.ex:240]: As mentioned in comment up, try to replace "bare" numbers with typed tensors

File: lib/scholar/impute/knn_imputer.ex [hunk 12]
```diff
@@ -0,0 +1,256 @@
+defmodule Scholar.Impute.KNNImputer do
```
Comment (msluszniak) [inline lib/scholar/impute/knn_imputer.ex:1]: I think it should be written with double t `KNNImputter` like formatter etc.

**Review by msluszniak — COMMENTED** [review]
Thanks for the PR, I dropped some comments :))

**Comment by krstopro** [comment]
Hi @srzeszut and thanks for the pull request. I’m traveling now and don’t have my laptop with me. Will be back this Sunday, so I will have a look probably next week.

File: lib/scholar/impute/knn_imputer.ex (hunk 2 above)
Comment (srzeszut) [reply lib/scholar/impute/knn_imputer.ex:80]: I found this check in simple imputer 
https://github.com/elixir-nx/scholar/blob/main/lib/scholar/impute/simple_imputer.ex
Are you sure it won't work?

File: lib/scholar/impute/knn_imputer.ex (hunk 5 above)
Comment (srzeszut) [reply lib/scholar/impute/knn_imputer.ex:158]: when i pass it like this  calculate_knn(x, row, col, rows, num_neighbors) the top_k function gets tensor instead of scalar and it does't work.

File: lib/scholar/impute/knn_imputer.ex (hunk 7 above)
Comment (srzeszut) [reply lib/scholar/impute/knn_imputer.ex:17]: could you suggest how to change it please?
because this is not compiling 
type: {:or, [:float, :integer,:nan]},

File: lib/scholar/impute/knn_imputer.ex [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+
+  # nan_col is the column of the value to impute
+  defnp nan_euclidian(row, nan_col, potential_neighbor) do
+    {coordinates} = Nx.shape(row)
+
+    # minus nan column
+    coordinates = coordinates - 1
+
+    # inputes zeros in nan_col to calculate distance with squared_euclidean
+    new_row = Nx.indexed_put(row, Nx.new_axis(nan_col, 0), Nx.tensor(0))
```
[the remark this answers is not in the document]
Comment (srzeszut) [reply lib/scholar/impute/knn_imputer.ex]: I changed it but I don't know how to change this line 
    row_distances = Nx.iota({rows}, type: {:f, 32})
because i don't know what the type calculated distance will be at this point

**Comment by srzeszut** [comment]
thanks for the review I apply suggested changes and leave some comments

File: lib/scholar/impute/knn_imputer.ex (hunk 2 above)
Comment (josevalim) [reply lib/scholar/impute/knn_imputer.ex]: It is also broken there. :)

File: lib/scholar/impute/knn_imputer.ex (hunk 7 above)
Comment (josevalim) [reply lib/scholar/impute/knn_imputer.ex]: This is correct as is. float and integer are types, nan is a value. :)

File: lib/scholar/impute/knn_imputter.ex [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+    row_nan_count = Nx.sum(Nx.is_nan(x), axes: [1])
+    # row with only 1 non nan value is also considered as all nan row
+    all_nan_rows =
+      Nx.select(Nx.greater_equal(row_nan_count, cols - 1), Nx.tensor(1), Nx.tensor(0))
+
+    all_nan_rows_count = Nx.sum(all_nan_rows)
+
+    if num_neighbors > rows - 1 - Nx.to_number(all_nan_rows_count) do
+      raise ArgumentError,
+            "Number of neighbors rows must be less than number valid of rows - 1 (valid row is row with more than 1 non nan value)"
```
Comment (josevalim) [inline lib/scholar/impute/knn_imputter.ex:104]: error messages start in lowercase. :)

```suggestion
            "number of neighbors rows must be less than number valid of rows - 1 (valid row is row with more than 1 non nan value)"
```

File: lib/scholar/impute/knn_imputter.ex [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+    {rows, cols} = Nx.shape(x)
+
+    row_nan_count = Nx.sum(Nx.is_nan(x), axes: [1])
+    # row with only 1 non nan value is also considered as all nan row
+    all_nan_rows =
+      Nx.select(Nx.greater_equal(row_nan_count, cols - 1), Nx.tensor(1), Nx.tensor(0))
+
+    all_nan_rows_count = Nx.sum(all_nan_rows)
+
+    if num_neighbors > rows - 1 - Nx.to_number(all_nan_rows_count) do
```
Comment (josevalim) [inline lib/scholar/impute/knn_imputter.ex:102]: Can you please add some tests? In particular, please add a test where you call jit this function and then you call it: `Nx.Defn.jit(...).(arg1, arg2)`. It should reveal some errors around here. :)

File: lib/scholar/impute/knn_imputer.ex (hunk 2 above)
Comment (josevalim) [reply lib/scholar/impute/knn_imputer.ex]: I have fixed it there: c024c5bc18dbbce4c1365adbbbb9cc26fa2738c5

File: lib/scholar/impute/knn_imputter.ex (hunk 15 above)
Comment (srzeszut) [reply lib/scholar/impute/knn_imputter.ex]: I added tests and checked it. I removed those checks and added them in the description

File: lib/scholar/impute/knn_imputter.ex [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+  ]
+
+  @opts_schema NimbleOptions.new!(opts_schema)
+
+  @doc """
+  Imputter for completing missing values using k-Nearest Neighbors.
+
+  Preconditions:
+    * `number_of_neighbors` is a positive integer.
+    *  number of neighbors must be less than number valid of rows - 1 (valid row is row with more than 1 non nan value) otherwise it is better to use simple imputter
```
Comment (josevalim) [inline lib/scholar/impute/knn_imputter.ex:39]: Please try to break this long line :)

File: lib/scholar/impute/knn_imputter.ex [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+    end
+
+    x =
+      if opts[:missing_values] != :nan,
+        do: Nx.select(Nx.equal(x, opts[:missing_values]), Nx.Constants.nan(), x),
+        else: x
+
+    num_neighbors = opts[:number_of_neighbors]
+
+    placeholder_value = Nx.Constants.nan() |> Nx.tensor()
```
Comment (josevalim) [inline lib/scholar/impute/knn_imputter.ex:90]: Can we move these three lines inside `knn_impute`, so it is all in `defn`? Btw, you can create a nan tensor with `Nx.tensor(:nan)`. :)

File: test/scholar/impute/knn_imputter_test.exs [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+      x = Nx.tensor([1, 2, 2, 3])
+
+      assert_raise ArgumentError,
+                   "Wrong input rank. Expected: 2, got: 1",
+                   fn ->
+                     KNNImputter.fit(x, missing_values: 1, number_of_neighbors: 2)
+                   end
+    end
+
+    test "Invalid n_neighbors value" do
```
Comment (josevalim) [inline test/scholar/impute/knn_imputter_test.exs:116]: Test names start in lowercase :)

```suggestion
    test "invalid impute rank" do
      x = Nx.tensor([1, 2, 2, 3])

      assert_raise ArgumentError,
                   "Wrong input rank. Expected: 2, got: 1",
                   fn ->
                     KNNImputter.fit(x, missing_values: 1, number_of_neighbors: 2)
                   end
    end

    test "invalid n_neighbors value" do
```

File: lib/scholar/impute/knn_imputter.ex [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+    missing_values: [
+      type: {:or, [:float, :integer, {:in, [:nan]}]},
+      default: :nan,
+      doc: ~S"""
+      The placeholder for the missing values. All occurrences of `:missing_values` will be imputed.
+
+      The default value expects there are no NaNs in the input tensor.
+      """
+    ],
+    number_of_neighbors: [
```
Comment (krstopro) [inline lib/scholar/impute/knn_imputter.ex:25]: I would suggest changing this to `num_neighbors` to be consistent with the rest of `Scholar`.

**Review by krstopro — COMMENTED** [review]
First review. Some features we might wanna have:
- Make k-NN algorithm configurable.
- Make the metric configurable.

You can leave these for another pull request. Have a look at e.g. `KNNClassifier` how it is done over there.
I should have another look tonight.

File: lib/scholar/impute/knn_imputter.ex (hunk 17 above)
Comment (srzeszut) [reply lib/scholar/impute/knn_imputter.ex:90]: when I moved  it it looks like Nx.equal(x, opts[:missing_values]) this is not working properly because
it changes everything to nan

Comment (josevalim) [reply lib/scholar/impute/knn_imputter.ex:90]: cool, thank you!

**Review by krstopro — COMMENTED** [review]
Several minor comments for now. I have to go through the code at least once more as I don't exactly understand the logic here.

File: lib/scholar/impute/knn_imputter.ex (hunk 16 above)
Comment (polvalente) [inline lib/scholar/impute/knn_imputter.ex:39]:
```suggestion
    *  The number of neighbors must be less than the number of valid rows - 1.
       A valid row is a row with more than 1 non-NaN values. Otherwise it is better to use a simpler imputer.
```

File: lib/scholar/impute/knn_imputter.ex [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return Values
+
+    The function returns a struct with the following parameters:
+
+    * `:missing_values` - the same value as in `:missing_values`
+
+    * `:statistics` - The imputation fill value for each feature. Computing statistics can result in
+    [`Nx.Constant.nan/0`](https://hexdocs.pm/nx/Nx.Constants.html#nan/0) values.
```
Comment (polvalente) [inline lib/scholar/impute/knn_imputter.ex:53]:
```suggestion
    [`Nx.Constants.nan/0`](https://hexdocs.pm/nx/Nx.Constants.html#nan/0) values.
```

Comment (polvalente) [reply lib/scholar/impute/knn_imputter.ex:53]: Do you need the explicit linking in hexdoc?

File: lib/scholar/impute/knn_imputter.ex [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return Values
+
+    The function returns a struct with the following parameters:
+
+    * `:missing_values` - the same value as in `:missing_values`
```
Comment (polvalente) [inline lib/scholar/impute/knn_imputter.ex:50]:
```suggestion
    * `:missing_values` - the same value as in the `:missing_values` option
```

File: lib/scholar/impute/knn_imputter.ex [hunk 22]
```diff
@@ -0,0 +1,247 @@
...
+
+    input_rank = Nx.rank(x)
+
+    if input_rank != 2 do
+      raise ArgumentError, "Wrong input rank. Expected: 2, got: #{inspect(input_rank)}"
+    end
+
+    x =
+      if opts[:missing_values] != :nan,
+        do: Nx.select(Nx.equal(x, opts[:missing_values]), Nx.Constants.nan(), x),
```
[the remark this answers is not in the document]
Comment (polvalente) [reply lib/scholar/impute/knn_imputter.ex:85]: This is a deftransform, so `Nx.equal` is the proper function. `==` will be `Elixir.Kernel.==`

File: lib/scholar/impute/knn_imputter.ex (hunk 17 above)
Comment (polvalente) [inline lib/scholar/impute/knn_imputter.ex:90]:
```suggestion
    placeholder_value = Nx.Constants.nan()
```

Comment (polvalente) [reply lib/scholar/impute/knn_imputter.ex:90]: you probably want to pass the input type here to avoid upcasts

File: lib/scholar/impute/knn_imputter.ex [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+  """
+  import Nx.Defn
+  import Scholar.Metrics.Distance
+
+  @derive {Nx.Container, keep: [:missing_values], containers: [:statistics]}
+  defstruct [:statistics, :missing_values]
+
+  opts_schema = [
+    missing_values: [
+      type: {:or, [:float, :integer, {:in, [:nan]}]},
```
Comment (polvalente) [inline lib/scholar/impute/knn_imputter.ex:17]:
```suggestion
      type: {:or, [:float, :integer, {:in, [:nan]}]},
```

I believe this should allow :infinity and :neg_infinity too for completeness

File: lib/scholar/impute/knn_imputter.ex [hunk 24]
```diff
@@ -0,0 +1,247 @@
...
+            if mask[row][col] > 0 do
+              {rows, cols} = Nx.shape(x)
+
+              neighbor_avg =
+                calculate_knn(x, row, col, rows: rows, num_neighbors: opts[:num_neighbors])
+
+              indices =
+                [Nx.stack(row), Nx.stack(col)]
+                |> Nx.concatenate()
+                |> Nx.stack()
```
Comment (polvalente) [inline lib/scholar/impute/knn_imputter.ex:146]:
```suggestion
              indices = Nx.stack([row, col]) |> Nx.reshape({1, 2})
```

If I read the code correctly, row and col are scalars and this should yield the same result

File: lib/scholar/impute/knn_imputter.ex [hunk 25]
```diff
@@ -0,0 +1,247 @@
...
+
+              values_to_impute = Nx.indexed_put(values_to_impute, indices, Nx.stack(neighbor_avg))
```
Comment (polvalente) [inline lib/scholar/impute/knn_imputter.ex:148]:
```suggestion
              values_to_impute = Nx.put_slice(values_to_impute, [row, col], Nx.reshape(neighbor_avg, {1, 1}))
```

I think this is even simpler

**Comment by srzeszut** [comment]
Thanks for all the comments, I applied your suggested changes to the code.

**Review by josevalim — APPROVED** [review]

**Review by msluszniak — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
